docs/SOFTWARE_DESIGN.md

```diff
@@ -0,0 +1,124 @@
+
+There is a very clear separation between front- and back-ends, so we can delegate disjoint tasks and responsibilities. The back-end game logic can be written in C++ without any knowledge of JS (and could ideally be modular enough to couple with a command-line front-end that might be useful for development). Greg will mostly be responsible for linking the JS and C++ components together.
+
+```
+           BACKEND (game engine)
+
+       .-------.
+       | uTest |
+       `-------`
+           |
+           v        [emscripten compilation]
+.-----------------.          |
+| C++ game engine |----------+--+
+`-----------------`             |
+                                |
+................................|.............
+                                |
+.------.   FRONTEND (browser)   |
+| User |                        |
+`------`                        |
+   |                            |
+[click events]    [API calls]   |
+   |                   |        v
+   |   .----------.~~~~+~~~>.------.
+   +-->| HTML5 UI |         | WASM |<--compiled backend engine
+       `----------`<~~~+~~~~`------`
+            ^          |
+            |     [game state]
+            |
+            |
+.----------------------.
+| Jest/Puppeteer tests |
+`----------------------`
+```
+
+Greg has built a simple proof-of-concept that illustrates the concept/workflow and relevant technologies interoperating: <https://github.com/ggorlen/wasm-click-counter>.
+
+### Game engine back-end
+
+The Sokoban game engine will consist of a top-level C++ class, `Sokoban`, which will expose the following interface:
+
+```
+// one of the possible objects that can exist on a cell
+enum Cell {
+   ' ': empty space
+   '#': impassable wall
+   '.': goal cell
+   '$': box
+   '*': box on top of goal cell
+   '@': player
+   '+': player on top of goal cell
+}
+
+// one of the possible movement directions for the player
+enum Direction {UP, DOWN, LEFT, RIGHT}
+
+// attempt a player move in direction (up/down/left/right) and return true/false to indicate if it was successful
+bool move(Direction direction)
+
+// switch to level level_number if possible, throw otherwise
+void change_level(int level_number)
+
+// reset the current level to its original state (could be replaced by change_level by passing in the current level number)
+void reset()
+
+// undo the last move
+void undo()
+
+// return the current board position as a 2d vector of characters
+vector<vector<Cell>> board()
+
+// return the number of moves made on the current board
+int moves()
+
+// return the current level number
+int level()
+
+// return whether the current board is solved
+bool solved()
+```
+
+Internally, levels will be a `vector<vector<vector<Cell>>>`, that is, a vector of sequential levels, each of which is a 2d vector representing the game board, with each cell being a character.
+
```

Comment:
Will the sequential levels be saved in a Directory containing files to corresponds to each level's map. Maybe we can save this in a hash table where the key is the level number and the value would be the location of the file we could read the map from.